File: backend/src/middleware/metrics.ts
/**
 * Prometheus Metrics Middleware
 * Collects HTTP and database metrics for monitoring
 */

import { Registry, Counter, Histogram, Gauge, collectDefaultMetrics } from 'prom-client';
import { Request, Response, NextFunction } from 'express';

// Create a metrics registry
export const register = new Registry();

// Collect default Node.js process metrics (CPU, memory, event loop, etc.)
collectDefaultMetrics({ register });

// HTTP Metrics
export const httpRequestsTotal = new Counter({
  name: 'http_requests_total',
  help: 'Total number of HTTP requests',
  labelNames: ['method', 'route', 'status_code'],
  registers: [register],
});

export const httpRequestDurationSeconds = new Histogram({
  name: 'http_request_duration_seconds',
  help: 'Duration of HTTP requests in seconds',
  labelNames: ['method', 'route', 'status_code'],
  buckets: [0.1, 0.5, 1, 2, 5, 10, 30],
  registers: [register],
});

export const httpRequestsInProgress = new Gauge({
  name: 'http_requests_in_progress',
  help: 'Number of HTTP requests currently in progress',
  registers: [register],
});

// Database Metrics
export const dbQueryDurationSeconds = new Histogram({
  name: 'db_query_duration_seconds',
  help: 'Duration of database queries in seconds',
  labelNames: ['query_type'],
  buckets: [0.01, 0.05, 0.1, 0.5, 1, 2, 5],
  registers: [register],
});

export const dbQueryErrorsTotal = new Counter({
  name: 'db_query_errors_total',
  help: 'Total number of database query errors',
  labelNames: ['query_type'],
  registers: [register],
});

/**
 * Middleware to instrument HTTP requests
 */
/**
 * Sanitize route path to reduce cardinality
 * Replaces UUIDs and IDs with placeholders
 */
function sanitizeRoute(route: string): string {
  if (!route) return 'unknown';
  
  // Replace UUIDs (8-4-4-4-12 format)
  route = route.replace(/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/gi, ':id');
  
  // Replace numeric IDs
  route = route.replace(/\/\d+/g, '/:id');
  
  // Replace long alphanumeric IDs
  route = route.replace(/\/[a-z0-9]{20,}/gi, '/:id');
  
  return route;
}

export function metricsMiddleware(req: Request, res: Response, next: NextFunction): void {
  const startTime = Date.now();
  // Use matched route pattern if available, otherwise sanitize the path
  const route = req.route?.path || sanitizeRoute(req.path) || 'unknown';
  const method = req.method;

  // Increment in-progress requests
  httpRequestsInProgress.inc();

  // Track response finish
  res.on('finish', () => {
    const duration = (Date.now() - startTime) / 1000;
    const statusCode = res.statusCode.toString();

    // Record metrics
    httpRequestsTotal.inc({ method, route, status_code: statusCode });
    httpRequestDurationSeconds.observe({ method, route, status_code: statusCode }, duration);
    httpRequestsInProgress.dec();
  });

  next();
}

/**
 * Record database query duration
 */
export function recordDbQuery(duration: number, queryType: 'query' | 'procedure' | 'transaction'): void {
  dbQueryDurationSeconds.observe({ query_type: queryType }, duration);
}

/**
 * Record database query error
 */
export function recordDbError(queryType: 'query' | 'procedure' | 'transaction'): void {
  dbQueryErrorsTotal.inc({ query_type: queryType });
}

File: rl-service/api/metrics.py
"""
Prometheus Metrics Module for RL Service
Collects ML inference metrics for monitoring
"""

from prometheus_client import Counter, Histogram, Gauge, generate_latest, REGISTRY
import time
from typing import Optional

# Inference Metrics
inference_requests_total = Counter(
    'inference_requests_total',
    'Total number of inference requests',
    ['model', 'status']
)

inference_duration_seconds = Histogram(
    'inference_duration_seconds',
    'Duration of inference requests in seconds',
    ['model'],
    buckets=[0.01, 0.05, 0.1, 0.5, 1, 2, 5, 10]
)

inference_errors_total = Counter(
    'inference_errors_total',
    'Total number of inference errors',
    ['model', 'error_type']
)

model_loaded = Gauge(
    'model_loaded',
    'Whether model is loaded (1) or not (0)',
    ['model']
)

# Database Metrics
db_query_duration_seconds = Histogram(
    'db_query_duration_seconds',
    'Duration of database queries in seconds',
    ['query_type'],
    buckets=[0.01, 0.05, 0.1, 0.5, 1, 2, 5]
)

db_query_errors_total = Counter(
    'db_query_errors_total',
    'Total number of database query errors',
    ['query_type']
)


def record_inference(duration: float, model_name: str = 'default', status: str = 'success'):
    """
    Record inference metrics
    
    Args:
        duration: Inference duration in seconds
        model_name: Name of the model used
        status: Status of the inference ('success' or 'error')
    """
    inference_requests_total.labels(model=model_name, status=status).inc()
    inference_duration_seconds.labels(model=model_name).observe(duration)


def record_error(model_name: str = 'default', error_type: str = 'unknown'):
    """
    Record inference error
    
    Args:
        model_name: Name of the model used
        error_type: Type of error that occurred
    """
    inference_errors_total.labels(model=model_name, error_type=error_type).inc()


def set_model_loaded(value: int, model_name: str = 'default'):
    """
    Set model loaded status
    
    Args:
        value: 1 if loaded, 0 if not
        model_name: Name of the model
    """
    model_loaded.labels(model=model_name).set(value)


def record_db_query(duration: float, query_type: str = 'select'):
    """
    Record database query metrics
    
    Args:
        duration: Query duration in seconds
        query_type: Type of query ('select', 'insert', 'update', 'delete')
    """
    db_query_duration_seconds.labels(query_type=query_type).observe(duration)


def record_db_error(query_type: str = 'select'):
    """
    Record database query error
    
    Args:
        query_type: Type of query that failed
    """
    db_query_errors_total.labels(query_type=query_type).inc()


def generate_metrics():
    """
    Generate Prometheus-formatted metrics
    
    Returns:
        bytes: Prometheus metrics in text format
    """
    return generate_latest(REGISTRY)

